SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp

```diff
@@ -0,0 +1,110 @@
+#include "EngineClasses/SpatialVirtualWorkerTranslator.h"
+#include "EngineClasses/SpatialNetDriver.h"
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "Interop/SpatialSender.h"
+#include "Interop/SpatialStaticComponentView.h"
+#include "Schema/AuthorityIntent.h"
+#include "SpatialCommonTypes.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialLoadBalanceEnforcer);
+
+using namespace SpatialGDK;
+
+USpatialLoadBalanceEnforcer::USpatialLoadBalanceEnforcer(const FObjectInitializer& ObjectInitializer)
+	: Super(ObjectInitializer)
+	, NetDriver(nullptr)
+	, Sender(nullptr)
+	, VirtualWorkerTranslator(nullptr)
+{
+}
+
+void USpatialLoadBalanceEnforcer::Init(USpatialNetDriver* InNetDriver,
+	USpatialSender* InUpatialSender,
+	SpatialVirtualWorkerTranslator* InVirtualWorkerTranslator)
+{
+	NetDriver = InNetDriver;
+	Sender = InUpatialSender;
+	VirtualWorkerTranslator = InVirtualWorkerTranslator;
+}
+
+void USpatialLoadBalanceEnforcer::Tick()
+{
+	ProcessQueuedAclAssignmentRequests();
+}
+
+void USpatialLoadBalanceEnforcer::OnComponentUpdated(const Worker_ComponentUpdateOp& Op)
+{
+	check(NetDriver != nullptr)
+	check(NetDriver->StaticComponentView != nullptr)
+	if (Op.update.component_id == SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID &&
+		NetDriver->StaticComponentView->GetAuthority(Op.entity_id, SpatialConstants::ENTITY_ACL_COMPONENT_ID) == WORKER_AUTHORITY_AUTHORITATIVE)
+	{
+		QueueAclAssignmentRequest(Op.entity_id);
+	}
+}
+
+void USpatialLoadBalanceEnforcer::AuthorityChanged(const Worker_AuthorityChangeOp& AuthOp)
+{
+	if (AuthOp.component_id == SpatialConstants::ENTITY_ACL_COMPONENT_ID &&
+		AuthOp.authority == WORKER_AUTHORITY_AUTHORITATIVE)
+	{
+		// We have gained authority over the ACL component for some entity
+		// If we have authority, the responsibility here is to set the EntityACL write auth to match the worker requested via the virtual worker component
+		QueueAclAssignmentRequest(AuthOp.entity_id);
+	}
+}
+
+void USpatialLoadBalanceEnforcer::QueueAclAssignmentRequest(const Worker_EntityId EntityId)
+{
+	if (AclWriteAuthAssignmentRequests.ContainsByPredicate([EntityId](const WriteAuthAssignmentRequest& Request) { return Request.EntityId == EntityId; }))
+	{
+		UE_LOG(LogSpatialLoadBalanceEnforcer, Log, TEXT("(%s) Already has an ACL authority request for entity %lld"), *NetDriver->Connection->GetWorkerId(), EntityId);
+	}
+	else
+	{
+		UE_LOG(LogSpatialLoadBalanceEnforcer, Log, TEXT("(%s) Queueing ACL assignment request for %lld"), *NetDriver->Connection->GetWorkerId(), EntityId);
+		AclWriteAuthAssignmentRequests.Add(WriteAuthAssignmentRequest(EntityId));
+	}
+}
+
+void USpatialLoadBalanceEnforcer::ProcessQueuedAclAssignmentRequests()
+{
+	TArray<Worker_EntityId> CompletedRequests;
+	CompletedRequests.Reserve(AclWriteAuthAssignmentRequests.Num());
+
+	for (WriteAuthAssignmentRequest& Request : AclWriteAuthAssignmentRequests)
+	{
+		static const int16 MaxNumProcessAttempts = 5;
```

Comment:
why int 16?

Reply:
Because the variable we are compare against (Request.ProcessAttempts) it is an int16. Perhaps that should be an int8.